File: behavioral/command/task_scheduler/typescript/commands.ts
// --- Command Interface ---
export interface Command {
  execute(): Promise<void>; // Using Promise for potential async operations
}

// --- Receivers (Perform the actual work) ---
export class EmailService {
  public async sendEmail(
    recipient: string,
    subject: string,
    body: string
  ): Promise<void> {
    console.log(`Sending email to ${recipient} with subject '${subject}'...`);
    console.log(`Body: ${body}`);
    // Simulate async work
    await new Promise((resolve) => setTimeout(resolve, 50));
    console.log("Email sent.");
  }
}

export class ReportGenerator {
  public async generateReport(
    reportType: string,
    outputPath: string
  ): Promise<void> {
    console.log(`Generating ${reportType} report...`);
    console.log(`Saving report to ${outputPath}...`);
    // Simulate async work
    await new Promise((resolve) => setTimeout(resolve, 100));
    console.log("Report generated.");
  }
}

export class DatabaseService {
  public async runBackup(backupName: string): Promise<void> {
    console.log(`Starting database backup '${backupName}'...`);
    // Simulate async work
    await new Promise((resolve) => setTimeout(resolve, 150));
    console.log(`Database backup '${backupName}' completed.`);
  }
}

// --- Concrete Commands ---
export class SendEmailCommand implements Command {
  constructor(
    private service: EmailService,
    private recipient: string,
    private subject: string,
    private body: string
  ) {}

  public async execute(): Promise<void> {
    await this.service.sendEmail(this.recipient, this.subject, this.body);
  }

  // Optional: For logging/debugging purposes
  public getDescription(): string {
    return `SendEmailCommand(to: ${this.recipient}, subject: ${this.subject})`;
  }
}

export class GenerateReportCommand implements Command {
  constructor(
    private service: ReportGenerator,
    private reportType: string,
    private outputPath: string
  ) {}

  public async execute(): Promise<void> {
    await this.service.generateReport(this.reportType, this.outputPath);
  }

  public getDescription(): string {
    return `GenerateReportCommand(type: ${this.reportType}, path: ${this.outputPath})`;
  }
}

export class RunDatabaseBackupCommand implements Command {
  constructor(private service: DatabaseService, private backupName: string) {}

  public async execute(): Promise<void> {
    await this.service.runBackup(this.backupName);
  }

  public getDescription(): string {
    return `RunDatabaseBackupCommand(name: ${this.backupName})`;
  }
}

File: behavioral/command/task_scheduler/python/commands.py
from abc import ABC, abstractmethod

# --- Command Interface ---
class Command(ABC):
    """The Command interface declares a method for executing a command."""
    @abstractmethod
    def execute(self) -> None:
        pass

# --- Receivers (Perform the actual work) ---
class EmailService:
    """Knows how to send emails."""
    def send_email(self, recipient: str, subject: str, body: str) -> None:
        print(f"Sending email to {recipient} with subject '{subject}'...")
        print(f"Body: {body}")
        print("Email sent.")

class ReportGenerator:
    """Knows how to generate reports."""
    def generate_report(self, report_type: str, output_path: str) -> None:
        print(f"Generating {report_type} report...")
        print(f"Saving report to {output_path}...")
        print("Report generated.")

class DatabaseService:
    """Knows how to perform database operations."""
    def run_backup(self, backup_name: str) -> None:
        print(f"Starting database backup '{backup_name}'...")
        # Simulate work
        print(f"Database backup '{backup_name}' completed.")


# --- Concrete Commands ---
class SendEmailCommand(Command):
    """Concrete command to send an email."""
    def __init__(self, service: EmailService, recipient: str, subject: str, body: str):
        self._service = service
        self._recipient = recipient
        self._subject = subject
        self._body = body

    def execute(self) -> None:
        self._service.send_email(self._recipient, self._subject, self._body)

class GenerateReportCommand(Command):
    """Concrete command to generate a report."""
    def __init__(self, service: ReportGenerator, report_type: str, output_path: str):
        self._service = service
        self._report_type = report_type
        self._output_path = output_path

    def execute(self) -> None:
        self._service.generate_report(self._report_type, self._output_path)

class RunDatabaseBackupCommand(Command):
    """Concrete command to run a database backup."""
    def __init__(self, service: DatabaseService, backup_name: str):
        self._service = service
        self._backup_name = backup_name

    def execute(self) -> None:
        self._service.run_backup(self._backup_name)
